rust/lancedb/src/database/listing/manifest.rs

```diff
@@ -0,0 +1,3991 @@
+        let write_params = WriteParams {
+            mode: WriteMode::Create,
+            session: Some(self.session.clone()),
+            ..Default::default()
+        };
+
+        let insert_builder = InsertBuilder::new(&self.manifest_uri).with_params(&write_params);
+
+        let mut dataset = insert_builder
+            .execute_stream(Box::new(RecordBatchIterator::new(
+                vec![Ok(RecordBatch::new_empty(schema.clone()))],
+                schema.clone(),
+            )))
+            .await
+            .map_err(|e| Error::Lance { source: e })?;
+
+        if let Err(e) = Self::create_manifest_index_object_id_btree(&mut dataset).await {
+            log::warn!(
+                "Failed to create BTree index on object_id: {}. Query performance may be impacted.",
+                e
+            );
+        }
+
+        if let Err(e) = Self::create_manifest_index_object_type_bitmap(&mut dataset).await {
+            log::warn!(
+                "Failed to create Bitmap index on object_type: {}. Query performance may be impacted.",
+                e
+            );
+        }
+
+        if let Err(e) = Self::create_manifest_index_base_objects_label_list(&mut dataset).await {
+            log::warn!(
+                "Failed to create LabelList index on base_objects: {}. Query performance may be impacted.",
+                e
+            );
+        }
+
+        let result = DatasetConsistencyWrapper::new_latest(
+            dataset,
+            self.manifest_config.read_consistency_interval,
+        );
+        Ok(result)
+    }
+
+    async fn create_manifest_index_object_id_btree(dataset: &mut Dataset) -> lance::Result<()> {
+        let btree_params = ScalarIndexParams::default();
+        dataset
+            .create_index(
+                &["object_id"],
+                IndexType::BTree,
+                Some("object_id_btree".to_string()),
+                &btree_params,
+                true,
+            )
+            .await
+    }
+
+    async fn create_manifest_index_object_type_bitmap(dataset: &mut Dataset) -> lance::Result<()> {
+        let bitmap_params = ScalarIndexParams::default();
+        dataset
+            .create_index(
+                &["object_type"],
+                IndexType::Bitmap,
+                Some("object_type_bitmap".to_string()),
+                &bitmap_params,
+                true,
+            )
+            .await
+    }
+
+    async fn create_manifest_index_base_objects_label_list(
+        dataset: &mut Dataset,
+    ) -> lance::Result<()> {
+        let label_list_params = ScalarIndexParams::default();
+        dataset
+            .create_index(
+                &["base_objects"],
+                IndexType::LabelList,
+                Some("base_objects_label_list".to_string()),
+                &label_list_params,
+                true,
+            )
+            .await
+    }
+
+    async fn ensure_manifest_table_exists(&self) -> Result<()> {
```

Comment:
Not super important but why don't we just initialize the manifest dataset when the database is created?

Reply:
The reason I did it this way is because it is possible that user only has read access and cannot even create the manifest table, and then the whole thing is not useable. But yeah it seems like we can check that and just turn the manifest_enabled to false in that case at listing database. let me update it in that way so we can remove the mutex.